aggregator/aggregator.go

```diff
@@ -227,18 +270,31 @@ func (agg *aggregator) AddMetricWithPoliciesList(
 	return err
 }
 
-func (agg *aggregator) Close() error {
+func (agg *aggregator) Close(closeType CloseType) error {
 	agg.Lock()
-	defer agg.Unlock()
-
 	if agg.state != aggregatorOpen {
-		return errAggregatorIsNotOpenOrClosed
+		agg.Unlock()
+		return errAggregatorExpectOpen
+	}
+	if closeType == GracefulClose {
+		agg.state = aggregatorClosing
+		agg.Unlock()
+		ctx, cancel := context.WithTimeout(context.Background(), agg.gracefulCloseTimeout)
+		defer cancel()
+		if err := agg.electionManager.Resign(ctx); err != nil {
+			agg.Lock()
+			agg.state = aggregatorOpen
+			agg.Unlock()
```

Comment:
Unfortunately we can't hold the lock while resigning (which is what we get with `defer`) because resigning needs to confirm the successful transfer of leader role with kv and as a result may take a long time (by default it times out after 5 minutes) due to underlying network blip/etcd issue, causing metrics ingestion to be backed up due to the aggregator not able to acquire the read lock in `AddMetricListWithPolicies()`.